android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Node.kt

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (C) 2017. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.badoo.ribs.core
+
+import android.os.Bundle
+import android.os.Parcelable
+import android.support.annotation.CallSuper
+import android.support.annotation.MainThread
+import android.support.annotation.VisibleForTesting
+import android.util.SparseArray
+import android.view.ViewGroup
+import com.badoo.ribs.core.view.RibView
+import com.badoo.ribs.core.view.ViewFactory
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.uber.rib.util.RibRefWatcher
+import java.util.UUID
+import java.util.concurrent.CopyOnWriteArrayList
+
+/**
+ * Responsible for handling the addition and removal of child nodes.
+ **/
+open class Node<V : RibView>(
+    val forClass: Class<*>,
```

Comment:
I checked how it is used, looks nice.

But I found one problem, probably can be solved later, but anyway.
From my point of view there is everything fine if you will put into (Rib A) two same Ribs -> (Rib B) + (Rib B).
With such approach with class name it would be impossible to distinguish in which container should we put first and in which should we put second.

Reply:
Agree on this, it's better to use something more distinctive than class (id or smth?)

Reply:
@Nublo @ShikaSD Class here is only for 1. mainly deciding which type of rib should go into which view container when routing (since it's only the type, an automatically generated unique id doesn't make sense there) 2. debugging node structure.

We still have rib id and tag with unique values even for nodes of the same type. Do you see a case where we still need it in the above 2 scenarios?

Reply:
I suspect if you have 2 instances of the same RIB, which can be filled with different contents based on dependency, and they should go to different containers based on the contents, it won't work?
I can imagine some RIB with two containers, and children which take a user instance and show some info. If you want to have one of them shown in the container 1, but another in the container 2, it will be a bit annoying to handle.

Reply:
See latest changes, should solve the above.